README.md

# Unity Star Renderer

This code is used to render a 3D star field into you scene using the [HYG star database](https://github.com/astronexus/HYG-Database).

In this animation you can see Orion and Pleiades as we fly along at 1 lightyear per second! While flying between stars

![Orion](./Documentation/Orion.webp)

It is also simple to keep the sky rendered from a fixed vantage point. Whether that is from Earth/Sol or... Betelgeuse?

This package is only compatible with the Universal Render Pipeline (URP). Though it should be easy to port to the HDRP as well.

## Technical overview

Star size is tricky. It isn't based on distance, but relative magnitude (brightness). And relative magnitude varies depending
on your position in the star field. With about 120,000 stars, and the goal of flying at any speed performance is key. Relative magnitude
is recalculated in the graphics shader for each star every frame. Then, based on this magnitude a star size is chosen. This sounds compute
intensive - and it would be on a CPU. But in a shader it runs amazingly fast.

Each star is rendered near the camera's far clip plane, so there's no worries of other scene objects rendering behind the stars.

Because you can fly as fast as you want through the whole star field all ~120,000 stars need to be loaded up at all times. Even the dimmest
star will be bright if you're close enough to it! You'll be rendering around 240,000 very simple polygons separate into about 24 meshes. But
again, performance doesn't seem to be an issue - even if you're targeting a 500fps (something else in your game will likely be your bottleneck).

There are many small touches that make this star field look good. These include red/blue shift of each star (again using real data), and a subtle
twinkle. Later on these parameters could be made more configurable.

Unity Shader Graph is used to author the shaders. Was it easier than writing them in code? Nope - but hopefully it ends up being more maintainable
and easier for others to tweak.

## Usage for interstellar travel

Since this is a 3D star field you can fly from star to star seamlessly. Your position is specified by the
"ViewerPosition" on the main "GenerateStars" Script. Position is relative to the sun (Sol) and is measured in Parsecs.

## Usage as a Skybox

If you plan to use this as a Skybox from the perspective of Earth's surface it is possible to specify a Latitude, Longitude,
Day of year (0-365) and time (0-24). The stars will rotate accordingly with +Z being north and +X being east. Based on simple
testing star positions seem to be within 1-2 degrees of true value.

You'll also want to set "Viewer Position" to 0, 0, 0 (the position of our sun, Sol) and disable "Viewer to Camera Position" (so
you don't fly 1 parsec for every unit you move the camera).

In the editor it is possible to type a city name, press "Search City" and have the latitude / longitude filled automatically.
Note however format must be exactly city name optionally followed by a 2 character country code. For example "Tokyo", "Tokyo, JP",
"Sydney", "Sydney, AU" are all valid. After searching the country code will be added if missing (so you know whether you got the
correct result).

## Demo

[WebGL demo available on simmer.io](https://simmer.io/@tjbaron/star-renderer)

## Adding to your project

### As a read only package

1. In Unity, got to "Window > Package Manager".
2. Click "+" and choose "Add package from git URL..."
3. Paste "https://github.com/tjbaron/unity-star-renderer.git" and press "Add"

Note: If you have trouble, ensure that both git and git-lfs are installed (on Mac this can be done via HomeBrew).

### As a git submodule

1. Ensure your project is already a git repo
2. Go to your projects root folder on command line
3. Run `git submodule add https://github.com/tjbaron/unity-star-renderer.git Packages/com.baroncreations.star-renderer`

You should now have the submodule added to the directory `Packages/com.baroncreations.star-renderer`

## Viewing the demo scenes

If you took the read only approach you will need to copy the example scenes to your "Assets" folder before opening:

1. In Unity "Projects" go to "Packages > Star Renderer > Samples"
2. Drag the scenes to the "Assets" folder to copy them over
3. Double click the scenes in you "Assets" folder to open


CHANGELOG.md
# Changelog

## 0.1.0

- First version
- Render stars
- Allow fixed vantage
- Allow flying between stars
- Flying between stars demo scene

## 0.2.0

- Ability to set latitude, longitude, day and time and have sky rotated (roughly) correctly
- Earth perspective demo scene
- Updated project to be valid Unity Package


Samples/Autopilot.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Autopilot : MonoBehaviour
{
    public bool groundMode = false;

    public bool autoPilot = false;

    private string closestStarName = "";
    private string searchText = "";
    private Vector3 closestStarPosition = Vector3.zero;

    // Used to smoothly rotate towards the a star after search.
    private Quaternion startRotation = Quaternion.identity;
    private Quaternion targetRotation = Quaternion.identity;
    private float rotationTime = 5f;

    private string azalt = "";

    public StarRenderer starRendererScript;

    void Update()
    {
        // Fly us to selected star if autopilot enabled
        if (autoPilot && closestStarName != "")
        {
            var travelPath = closestStarPosition - Camera.main.transform.position;
            var travelVector = travelPath.normalized;
            var travelDistance = Mathf.Min(Time.deltaTime / 3.26156f, travelPath.magnitude); // 1 lightyear per second
            Camera.main.transform.position += (travelVector * travelDistance);
        }

        // Select a star by right clicking it
        if (Input.GetMouseButtonDown(1))
        {
            var starData = starRendererScript.starDataAsset.text;
            var closestStarDistance = 99999999999999f;
            var rows = starData.Split('\n');
            for (var i = 1; i < rows.Length - 1; i++)
            {
                var row = rows[i];
                var cols = row.Split(',');
                var x = float.Parse(cols[(int)StarData.x]);
                var y = float.Parse(cols[(int)StarData.z]);
                var z = float.Parse(cols[(int)StarData.y]);
                var mag = float.Parse(cols[(int)StarData.mag]);
                var starName = cols[(int)StarData.proper];
                var hip = cols[(int)StarData.hip];

                Vector3 starPosition = starRendererScript.GetSkyQuaternion() * new Vector3(x, y, z); //Camera.main.transform.position;
                float starDistanceFromCamera = Vector3.Distance(starPosition, Camera.main.transform.position);

                if (mag < 20 && starName != "" && starDistanceFromCamera > 1f)
                {
                    Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition);
                    
                    float distance = Vector3.Cross(ray.direction, starPosition - starRendererScript.viewerPosition).magnitude;
                    float angularDistance = distance / Mathf.Max(starDistanceFromCamera, 0.1f); // relative distance... no particular units

                    if (angularDistance < closestStarDistance)
                    {
                        closestStarName = starName + " (HIP " + hip + ")";
                        closestStarPosition = starPosition;
                        closestStarDistance = angularDistance;
                    }
                }
            }
        }

        if (rotationTime < 5) {
            if (Time.deltaTime < 0.2f) rotationTime += Time.deltaTime;
            transform.rotation = Quaternion.Slerp(startRotation, targetRotation, rotationTime / 5f);
        }
    }

    private void SearchStar() {
        var starData = starRendererScript.starDataAsset.text;
        var rows = starData.Split('\n');
        for (var i = 1; i < rows.Length - 1; i++)
        {
            var row = rows[i];
            var cols = row.Split(',');
            var x = float.Parse(cols[(int)StarData.x]);
            var y = float.Parse(cols[(int)StarData.z]);
            var z = float.Parse(cols[(int)StarData.y]);
            var mag = float.Parse(cols[(int)StarData.mag]);
            var starName = cols[(int)StarData.proper];
            var hip = cols[(int)StarData.hip];

            Vector3 starPosition = starRendererScript.GetSkyQuaternion() * new Vector3(x, y, z);

            if (starName == searchText || hip == searchText) {
                closestStarName = starName + " (HIP " + hip + ")";
                closestStarPosition = starPosition;

                startRotation = transform.rotation;
                targetRotation = Quaternion.LookRotation(starPosition);
                rotationTime = 0;

                var azAltVec = starRendererScript.GetAzAlt(new Vector3(x, y, z));
                azalt = azAltVec.x.ToString("0.0") + " / " + azAltVec.y.ToString("0.0");
            }
        }
    }

    private void OnGUI()
    {
        GUILayout.BeginHorizontal();

        if (!groundMode) {
            if (autoPilot)
            {
                GUI.color = Color.green;
            } 
            if (GUILayout.Button("Autopilot"))
            {
                autoPilot = !autoPilot;
            }
            GUI.color = Color.white;
        }

        searchText = GUILayout.TextField(searchText, GUILayout.Width(200));
        if (GUILayout.Button("Search")) {
            SearchStar();
        }

        GUILayout.EndHorizontal();

        if (groundMode) {
            GUILayout.Label("September 20 15:00GMT");
            GUILayout.BeginHorizontal();
            GUILayout.Label("Latitude: " + starRendererScript.longLat.y.ToString("0.000"));
            starRendererScript.longLat.y = GUILayout.HorizontalSlider(starRendererScript.longLat.y, 90, -90, GUILayout.Width(200));
            GUILayout.EndHorizontal();
            GUILayout.BeginHorizontal();
            GUILayout.Label("Longitude: " + starRendererScript.longLat.x.ToString("0.000"));
            starRendererScript.longLat.x = GUILayout.HorizontalSlider(starRendererScript.longLat.x, -180f, 180f, GUILayout.Width(200));
            GUILayout.EndHorizontal();
        }

        GUILayout.Label(closestStarName);
        if (!autoPilot)
        {
            var saveCam = Camera.main.transform.position;
            Camera.main.transform.position = starRendererScript.viewerPosition;
            var screenPosition = Camera.main.WorldToScreenPoint(closestStarPosition);
            Camera.main.transform.position = saveCam;

            if (screenPosition.z > 0)
            {
                GUI.Label(new Rect(screenPosition.x - 7, Screen.height - screenPosition.y - 7, 20, 20), "O");
            }
        }

        if (!groundMode) {
            if (closestStarName == "")
            {
                GUILayout.Label("Right click to select a star");
                GUILayout.Label("Click Autopilot to fly to it");
                GUILayout.Label("Search stars using their HIP number or proper name");
            } 
        } else {
            if (closestStarName == "")
            {
                GUILayout.Label("Search stars using their HIP number or proper name");
            } 
        }

        GUILayout.Label("Az / Alt: " + azalt);
    }
}


Editor/StarRendererEditor.cs
using System;
using UnityEditor;
using UnityEngine;
using UnityEngine.UIElements;

[CustomEditor(typeof(StarRenderer))]
public class StarRendererEditor : Editor
{

    private VisualElement viewerPosition;

    public override VisualElement CreateInspectorGUI()
    {
        var visualTree = AssetDatabase.LoadAssetAtPath<VisualTreeAsset>("Packages/com.baroncreations.star-renderer/Editor/StarRendererEditor.uxml");
        var uiDoc = visualTree.CloneTree();

        viewerPosition = uiDoc.Q<Vector3Field>("viewerPosition");
        uiDoc.Q<Toggle>("viewerToggle").RegisterValueChangedCallback((evt) => {
            viewerPosition.style.display = !evt.newValue ? DisplayStyle.Flex : DisplayStyle.None;
        });

        var cityName = uiDoc.Q<TextField>("cityName");
        var citySearch = uiDoc.Q<Button>("citySearch");
        var longLat = uiDoc.Q<Vector2Field>("longLat");
        citySearch.clicked += () => {
            var city = GetCity(cityName.value);
            if (city != null)
            {
                cityName.value = city.name + ", " + city.country;
                longLat.value = new Vector2(city.lng, city.lat);
            }
        };

        var nowButton = uiDoc.Q<Button>("nowButton");
        var dayOfYear = uiDoc.Q<IntegerField>("dayOfYear");
        var gmtTime = uiDoc.Q<FloatField>("gmtTime");
        nowButton.clicked += () => {
            var now = DateTime.UtcNow;
            dayOfYear.value = now.DayOfYear;
            gmtTime.value = now.Hour + now.Minute / 60f;
        };

        return uiDoc;
    }

    public City GetCity(string name)
    {
        var json = AssetDatabase.LoadAssetAtPath<TextAsset>("Packages/com.baroncreations.star-renderer/Editor/cities.json");
        var jsonText = "{\"cities\": " + json.text + "}";
        var cities = JsonUtility.FromJson<Cities>(jsonText);
        foreach (var city in cities.cities)
        {
            if ((city.name + ", " + city.country).ToLower() == name.ToLower() || city.name.ToLower() == name.ToLower())
            {
                return city;
            }
        }
        return null;
    }
}

[Serializable]
public class City
{
    public string country;
    public string name;
    public float lat;
    public float lng;
}

[Serializable]
public class Cities
{
    public City[] cities;
}


Runtime/Scripts/GenerateStars.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public enum StarData {
    id,hip,hd,hr,gl,bf,proper,ra,dec,
    dist,pmra,pmdec,rv,mag,absmag,spect,ci,
    x,y,z,vx,vy,vz,rarad,decrad,
    pmrarad,pmdecrad,bayer,flam,con,comp,comp_primary,
    _base,lum,var,var_min,var_max
}

[RequireComponent(typeof(MeshRenderer))]
[RequireComponent(typeof(MeshFilter))]
public class GenerateStars : MonoBehaviour
{
    // This expects the hygdata_v3.csv star data file
    // https://github.com/astronexus/HYG-Database
    public TextAsset starDataAsset;
    public Material starMaterial;
    // Viewer position in parsecs relative to earth
    public Vector3 viewerPosition;
    public bool viewerToCameraPosition = true;

    // Default to new years on the north pole
    public Vector2 longLat = new Vector2(0f, 90f);
    public int dayOfYear = 0;
    public float gmtTime = 0f;

    public Vector2 GetSkyRotation() {
        transform.rotation = Quaternion.identity;
        float longitude = longLat.x;
        float latitude = longLat.y;

        // transform.Rotate(0, (longitude+55) + ((gmtTime-15)*360/24) + (dayOfYear*360/365), 0, Space.World);
        // transform.Rotate(90-latitude, 0, 0, Space.World);
        var degRotations = new Vector2(
            (longitude+55) + ((gmtTime-15)*360/24) + (dayOfYear*360/365),
            90-latitude
        );
        return degRotations * Mathf.Deg2Rad;
    }

    // From perspective of earth surface
    // Z axis is North
    // X axis is East
    public Quaternion GetSkyQuaternion() {
        Vector2 lonLat = GetSkyRotation();
        var go = new GameObject();
        go.transform.Rotate(0, lonLat.x*Mathf.Rad2Deg, 0, Space.World);
        go.transform.Rotate(lonLat.y*Mathf.Rad2Deg, 0, 0, Space.World);
        var rot = go.transform.rotation;
        Destroy(go);
        return rot;
    }

    public Vector2 GetAzAlt(Vector3 starPosition) {
        var starPositionLocal = GetSkyQuaternion() * starPosition;
        var az  = Vector3.SignedAngle(new Vector3(starPositionLocal.x, 0f, starPositionLocal.z), Vector3.forward, -Vector3.up);
        var alt = 90f - Vector3.Angle(starPositionLocal, Vector3.up);
        if (az < 0) az += 360f;
        return new Vector2(az, alt);
    }

    //public string[] highlightedStars = new string[]{"Alnitak", "Alnilam", "Mintaka"};

    void Start()
    {
        transform.position = Vector3.zero;
        transform.rotation = Quaternion.identity;
        transform.localScale = Vector3.one;

        // Get all the stars from hyg vdatabase
        var starData = starDataAsset.text;
        var rows = starData.Split('\n');
        var starCount = rows.Length-2;
        var chunkSize = 5000;

        
        Mesh mesh = new Mesh();
        MeshFilter meshFilter = new MeshFilter();
        Vector3[] vertices = new Vector3[chunkSize*4];
        int[] tris = new int[chunkSize*6];
        Vector3[] normals = new Vector3[chunkSize*4];
        Vector2[] uv = new Vector2[chunkSize*4];
        Vector2[] uv2 = new Vector2[chunkSize*4];
        GameObject gameObject;
        for (var i=0; i<starCount; i++) {
            // Create a new mesh for each group of `chunkSize` stars
            if (i % chunkSize == 0) {
                if (i != 0) {
                    mesh.vertices = vertices;
                    mesh.triangles = tris;
                    mesh.normals = normals;
                    mesh.uv = uv;
                    mesh.uv2 = uv2;
                    meshFilter.mesh = mesh;
                }
                mesh = new Mesh();
                vertices = new Vector3[chunkSize*4];
                tris = new int[chunkSize*6];
                normals = new Vector3[chunkSize*4];
                uv = new Vector2[chunkSize*4];
                uv2 = new Vector2[chunkSize*4];
                gameObject = new GameObject();
                gameObject.name = "StarBatch";
                gameObject.transform.parent = this.gameObject.transform;
                meshFilter = gameObject.AddComponent<MeshFilter>();
                var meshRenderer = gameObject.AddComponent<MeshRenderer>();
                meshRenderer.material = starMaterial;
            }
            var j = i % chunkSize;

            // Create a quad for each star
            var cols = rows[i+1].Split(',');
            var x = float.Parse( cols[(int)StarData.x] );
            var y = float.Parse( cols[(int)StarData.z] );
            var z = float.Parse( cols[(int)StarData.y] );
            var absmag = float.Parse( cols[(int)StarData.absmag] );
            var colorindex = 0f;
            try
            {
                colorindex = float.Parse(cols[(int)StarData.ci]); // <0 is blue, >0 is red
            } catch { }

            vertices[j*4] = new Vector3(x, y, z);
            vertices[j*4 + 1] = new Vector3(x, y, z);
            vertices[j*4 + 2] = new Vector3(x, y, z);
            vertices[j*4 + 3] = new Vector3(x, y, z);

            tris[j*6] = j*4;
            tris[j*6 + 1] = j*4 + 2;
            tris[j*6 + 2] = j*4 + 1;
            tris[j*6 + 3] = j*4 + 2;
            tris[j*6 + 4] = j*4 + 3;
            tris[j*6 + 5] = j*4 + 1;

            normals[j*4] = -Vector3.forward;
            normals[j*4 + 1] = -Vector3.forward;
            normals[j*4 + 2] = -Vector3.forward;
            normals[j*4 + 3] = -Vector3.forward;

            //var highlighted = 0f;
            //for (var m=0; m<highlightedStars.Length; m++) {
            //    if (cols[(int)StarData.proper] == highlightedStars[m]) {
            //        highlighted = 1f;
            //        Debug.Log("Found " + highlightedStars[m]);
            //        break;
            //    }
            //}

            uv[j*4] = new Vector2(0f, 0f);
            uv[j*4 + 1] = new Vector2(1f, 0f);
            uv[j*4 + 2] = new Vector2(0f, 1f);
            uv[j*4 + 3] = new Vector2(1f, 1f);
            uv2[j*4] = uv2[j*4 + 1] = uv2[j*4 + 2] = uv2[j*4 + 3] = new Vector2(absmag, colorindex);
        }
        mesh.vertices = vertices;
        mesh.triangles = tris;
        mesh.normals = normals;
        mesh.uv = uv;
        mesh.uv2 = uv2;
        meshFilter.mesh = mesh;
    }

    void Update() {
        transform.position = Vector3.zero;
        transform.rotation = Quaternion.identity;
        transform.localScale = Vector3.one;
        Vector2 skyRotation = GetSkyRotation();

        if (viewerToCameraPosition)
        {
            viewerPosition = Camera.main.transform.position;
        }
        starMaterial.SetVector("ViewerPosition", viewerPosition);
        starMaterial.SetVector("CameraUp", Quaternion.Inverse(transform.rotation)*Camera.main.transform.up);
        starMaterial.SetVector("CameraRight", Quaternion.Inverse(transform.rotation)*Camera.main.transform.right);
        starMaterial.SetVector("_StarRotation", skyRotation);
    }

}


Runtime/Scripts/StarRenderer.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public enum StarData {
    id,hip,hd,hr,gl,bf,proper,ra,dec,
    dist,pmra,pmdec,rv,mag,absmag,spect,ci,
    x,y,z,vx,vy,vz,rarad,decrad,
    pmrarad,pmdecrad,bayer,flam,con,comp,comp_primary,
    _base,lum,var,var_min,var_max
}

[RequireComponent(typeof(MeshRenderer))]
[RequireComponent(typeof(MeshFilter))]
public class StarRenderer : MonoBehaviour
{
    // This expects the hygdata_v3.csv star data file
    // https://github.com/astronexus/HYG-Database
    public TextAsset starDataAsset;
    public Material starMaterial;
    // Viewer position in parsecs relative to earth
    public Vector3 viewerPosition;
    public bool viewerToCameraPosition = true;

    // Default to new years on the north pole
    public Vector2 longLat = new Vector2(0f, 90f);
    public int dayOfYear = 0;
    public float gmtTime = 0f;

    public Vector2 GetSkyRotation() {
        transform.rotation = Quaternion.identity;
        float longitude = longLat.x;
        float latitude = longLat.y;

        // transform.Rotate(0, (longitude+55) + ((gmtTime-15)*360/24) + (dayOfYear*360/365), 0, Space.World);
        // transform.Rotate(90-latitude, 0, 0, Space.World);
        var degRotations = new Vector2(
            (longitude+55) + ((gmtTime-15)*360/24) + (dayOfYear*360/365),
            90-latitude
        );
        return degRotations * Mathf.Deg2Rad;
    }

    // From perspective of earth surface
    // Z axis is North
    // X axis is East
    public Quaternion GetSkyQuaternion() {
        Vector2 lonLat = GetSkyRotation();
        var go = new GameObject();
        go.transform.Rotate(0, lonLat.x*Mathf.Rad2Deg, 0, Space.World);
        go.transform.Rotate(lonLat.y*Mathf.Rad2Deg, 0, 0, Space.World);
        var rot = go.transform.rotation;
        Destroy(go);
        return rot;
    }

    public Vector2 GetAzAlt(Vector3 starPosition) {
        var starPositionLocal = GetSkyQuaternion() * starPosition;
        var az  = Vector3.SignedAngle(new Vector3(starPositionLocal.x, 0f, starPositionLocal.z), Vector3.forward, -Vector3.up);
        var alt = 90f - Vector3.Angle(starPositionLocal, Vector3.up);
        if (az < 0) az += 360f;
        return new Vector2(az, alt);
    }

    //public string[] highlightedStars = new string[]{"Alnitak", "Alnilam", "Mintaka"};

    void Start()
    {
        if (!starDataAsset) {
            Debug.LogError("No 'starDataAsset' set on 'GenerateStars' script");
            transform.name = "[ERROR] " + transform.name;
            return;
        }
        if (!starMaterial) {
            Debug.LogError("No 'starMaterial' set on 'GenerateStars' script");
            transform.name = "[ERROR] " + transform.name;
            return;
        }

        transform.position = Vector3.zero;
        transform.rotation = Quaternion.identity;
        transform.localScale = Vector3.one;

        // Get all the stars from hyg vdatabase
        var starData = starDataAsset.text;
        var rows = starData.Split('\n');
        var starCount = rows.Length-2;
        var chunkSize = 5000;

        
        Mesh mesh = new Mesh();
        MeshFilter meshFilter = new MeshFilter();
        Vector3[] vertices = new Vector3[chunkSize*4];
        int[] tris = new int[chunkSize*6];
        Vector3[] normals = new Vector3[chunkSize*4];
        Vector2[] uv = new Vector2[chunkSize*4];
        Vector2[] uv2 = new Vector2[chunkSize*4];
        GameObject gameObject;
        for (var i=0; i<starCount; i++) {
            // Create a new mesh for each group of `chunkSize` stars
            if (i % chunkSize == 0) {
                if (i != 0) {
                    mesh.vertices = vertices;
                    mesh.triangles = tris;
                    mesh.normals = normals;
                    mesh.uv = uv;
                    mesh.uv2 = uv2;
                    meshFilter.mesh = mesh;
                }
                mesh = new Mesh();
                vertices = new Vector3[chunkSize*4];
                tris = new int[chunkSize*6];
                normals = new Vector3[chunkSize*4];
                uv = new Vector2[chunkSize*4];
                uv2 = new Vector2[chunkSize*4];
                gameObject = new GameObject();
                gameObject.name = "StarBatch";
                gameObject.transform.parent = this.gameObject.transform;
                meshFilter = gameObject.AddComponent<MeshFilter>();
                var meshRenderer = gameObject.AddComponent<MeshRenderer>();
                meshRenderer.material = starMaterial;
            }
            var j = i % chunkSize;

            // Create a quad for each star
            var cols = rows[i+1].Split(',');
            var x = float.Parse( cols[(int)StarData.x] );
            var y = float.Parse( cols[(int)StarData.z] );
            var z = float.Parse( cols[(int)StarData.y] );
            var absmag = float.Parse( cols[(int)StarData.absmag] );
            var colorindex = 0f;
            try
            {
                colorindex = float.Parse(cols[(int)StarData.ci]); // <0 is blue, >0 is red
            } catch { }

            vertices[j*4] = new Vector3(x, y, z);
            vertices[j*4 + 1] = new Vector3(x, y, z);
            vertices[j*4 + 2] = new Vector3(x, y, z);
            vertices[j*4 + 3] = new Vector3(x, y, z);

            tris[j*6] = j*4;
            tris[j*6 + 1] = j*4 + 2;
            tris[j*6 + 2] = j*4 + 1;
            tris[j*6 + 3] = j*4 + 2;
            tris[j*6 + 4] = j*4 + 3;
            tris[j*6 + 5] = j*4 + 1;

            normals[j*4] = -Vector3.forward;
            normals[j*4 + 1] = -Vector3.forward;
            normals[j*4 + 2] = -Vector3.forward;
            normals[j*4 + 3] = -Vector3.forward;

            //var highlighted = 0f;
            //for (var m=0; m<highlightedStars.Length; m++) {
            //    if (cols[(int)StarData.proper] == highlightedStars[m]) {
            //        highlighted = 1f;
            //        Debug.Log("Found " + highlightedStars[m]);
            //        break;
            //    }
            //}

            uv[j*4] = new Vector2(0f, 0f);
            uv[j*4 + 1] = new Vector2(1f, 0f);
            uv[j*4 + 2] = new Vector2(0f, 1f);
            uv[j*4 + 3] = new Vector2(1f, 1f);
            uv2[j*4] = uv2[j*4 + 1] = uv2[j*4 + 2] = uv2[j*4 + 3] = new Vector2(absmag, colorindex);
        }
        mesh.vertices = vertices;
        mesh.triangles = tris;
        mesh.normals = normals;
        mesh.uv = uv;
        mesh.uv2 = uv2;
        meshFilter.mesh = mesh;
    }

    void Update() {
        if (!starMaterial) {
            return;
        }

        transform.position = Vector3.zero;
        transform.rotation = Quaternion.identity;
        transform.localScale = Vector3.one;
        Vector2 skyRotation = GetSkyRotation();

        if (viewerToCameraPosition)
        {
            viewerPosition = Camera.main.transform.position;
        }
        starMaterial.SetVector("ViewerPosition", viewerPosition);
        starMaterial.SetVector("CameraUp", Quaternion.Inverse(transform.rotation)*Camera.main.transform.up);
        starMaterial.SetVector("CameraRight", Quaternion.Inverse(transform.rotation)*Camera.main.transform.right);
        starMaterial.SetVector("_StarRotation", skyRotation);
    }

}
